presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSourceBase.java

```diff
@@ -0,0 +1,384 @@
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.common.type.VarcharType;
+import com.facebook.presto.pinot.query.PinotQueryGenerator;
+import com.facebook.presto.spi.ConnectorPageSource;
+import com.facebook.presto.spi.ConnectorSession;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.pinot.PinotErrorCode.PINOT_DECODE_ERROR;
+import static com.facebook.presto.pinot.PinotErrorCode.PINOT_EXCEPTION;
+import static com.facebook.presto.pinot.PinotErrorCode.PINOT_INSUFFICIENT_SERVER_RESPONSE;
+import static com.facebook.presto.pinot.PinotErrorCode.PINOT_UNEXPECTED_RESPONSE;
+import static com.facebook.presto.pinot.PinotErrorCode.PINOT_UNSUPPORTED_COLUMN_TYPE;
+import static com.facebook.presto.pinot.PinotUtils.doWithRetries;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.Boolean.parseBoolean;
+import static java.lang.Long.parseLong;
+import static java.util.Objects.requireNonNull;
+
+public abstract class PinotBrokerPageSourceBase
+        implements ConnectorPageSource
+{
+    private static final String PINOT_INFINITY = "∞";
+    private static final String PINOT_POSITIVE_INFINITY = "+" + PINOT_INFINITY;
+    private static final String PINOT_NEGATIVE_INFINITY = "-" + PINOT_INFINITY;
+
+    private static final Double PRESTO_INFINITY = Double.POSITIVE_INFINITY;
+    private static final Double PRESTO_NEGATIVE_INFINITY = Double.NEGATIVE_INFINITY;
+
+    protected final PinotConfig pinotConfig;
+    protected final List<PinotColumnHandle> columnHandles;
+    protected final List<PinotColumnHandle> expectedHandles;
+    protected final PinotClusterInfoFetcher clusterInfoFetcher;
+    protected final ConnectorSession session;
+    protected final ObjectMapper objectMapper;
+
+    protected boolean finished;
+    protected long readTimeNanos;
+    protected long completedBytes;
+
+    public PinotBrokerPageSourceBase(
+            PinotConfig pinotConfig,
+            ConnectorSession session,
+            List<PinotColumnHandle> columnHandles,
+            List<PinotColumnHandle> expectedHandles,
+            PinotClusterInfoFetcher clusterInfoFetcher,
+            ObjectMapper objectMapper)
+    {
+        this.pinotConfig = requireNonNull(pinotConfig, "pinot config is null");
+        this.clusterInfoFetcher = requireNonNull(clusterInfoFetcher, "cluster info fetcher is null");
+        this.columnHandles = ImmutableList.copyOf(columnHandles);
+        this.expectedHandles = requireNonNull(expectedHandles, "expected handles is null");
```

Comment:
Take a look at the comment below: expectedHandles may not even belong here.

Reply:
moved